src/Interfaces/hamiltonians.jl

```
@@ -440,54 +449,197 @@ struct OffdiagonalsOperatorColumn{A,T,O<:Union{AbstractOperator{T},AbstractMatri
+
+See also [`operator_column`](@ref), [`AbstractOperatorColumn`](@ref),
+and [`AbstractDVec`](@ref).
+"""
+function LinearAlgebra.dot(
+    dv, column::AbstractOperatorColumn
+)
+    res = conj(dv[starting_address(column)]) * diagonal_element(column)
+    for (newaddress, matrixelement) in offdiagonals(column)
+        res += conj(dv[newaddress]) * matrixelement
+    end
+    return res
+end
+function LinearAlgebra.dot(column::AbstractOperatorColumn, a)
+    return adjoint(LinearAlgebra.dot(a, column))
+end
+
+"""
+    offdiagonals(column::AbstractOperatorColumn)
+    offdiagonals(h::AbstractHamiltonian, address) # (deprecated)
+
+Return an iterator over nonzero off-diagonal matrix elements of `h` in the same column as
+`address`. Will iterate over pairs `(newaddress, matrixelement)` or
+`newaddress => matrixelement`.
+
+# Example
+
+```jldoctest
+julia> address = BoseFS(3,2,1);
+
+julia> H = HubbardReal1D(address);
+
+julia> h = offdiagonals(H * address)
+6-element Rimu.Hamiltonians.Offdiagonals{BoseFS{6, 3, BitString{8, 1, UInt8}}, Float64, HubbardReal1D{Float64, BoseFS{6, 3, BitString{8, 1, UInt8}}, 1.0, 1.0}}:
+ (fs"|2 3 1⟩", -3.0)
+ (fs"|2 2 2⟩", -2.449489742783178)
+ (fs"|3 1 2⟩", -2.0)
+ (fs"|4 1 1⟩", -2.8284271247461903)
+ (fs"|4 2 0⟩", -2.0)
+ (fs"|3 3 0⟩", -1.7320508075688772)
+```
+Part of the [`AbstractHamiltonian`](@ref) interface. See also
+[`AbstractOperatorColumn`](@ref) and [`operator_column`](@ref).
+"""
+function offdiagonals(m::AbstractMatrix, i)
+    pairs = collect(zip(axes(m, 1), view(m, :, i)))
+    return filter!(pairs) do ((k, v))
+        k ≠ i && v ≠ 0
+    end
+end
+function offdiagonals(::AbstractOperatorColumn{A,T,O}) where {A,T,O}
+    if has_iterable_offdiagonals(O)
+        error("offdiagonals not implemented for operator type $O " *
+              "even though has_iterable_offdiagonals($O) == true")
+    else
+        throw(ArgumentError("offdiagonals not supported for operator type $O"))
+    end
+end
```

I'd prefer if we just let it throw a `MethodError` if offdiagonals are not defined. Similar for `random_offdiagonal` below.